docs/proposals/24-dynamic-node-groups-for-shoot-cluster-autoscaler.md

```diff
@@ -0,0 +1,338 @@
+				  MaxUnavailable:       worker.DistributePositiveIntOrPercent(zoneIdx, pool.MaxUnavailable, zoneLen, pool.Minimum),
+				  Labels:               addTopologyLabel(pool.Labels, zone),
+				  Annotations:          pool.Annotations,
+          //...
+			})
+    }
+}
+```
+
+##### MachineDeployment Min/Max Distribution 
+
+As can be seen from the snippet above, the machine deployment `Minimum` and `Maximum` delegates to the helper method `DistributeOverZones` which is coded as below:
+
+```go
+func DistributeOverZones(zoneIndex, size, zoneSize int32) int32 {
+	first := size / zoneSize
+	second := int32(0)
+	if zoneIndex < (size % zoneSize) {
+		second = 1
+	}
+	return first + second
+}
+```
+Even machine deployment `MaxSurge` and `MaxUnavailable` delegates to `DistributePositiveIntOrPercent` which in turn delegates to `DistributeOverZones`
+
+The above computation has issues as can be seen below for generated machine deployments.
+
+###### Pool-0:1:1:1:2 (min:max:maxsurge:maxunavail:numzones)
+```
+MachineDeployment_Z0(Minimum=0|Maximum:1|MaxSurge:1|MaxUnavailable:1)
+MachineDeployment_Z1(Minimum=0|Maximum:0|MaxSurge:0|MaxUnavailable:0)
+```
+Consider the case above where `numzones:2` > worker pool `max:1`, We can see that the machinedeployment for `Z1` is useless as its `Min`, `Max`, `MaxSurge` and `MaxUnavailable` are all zero making the corresponding CA `NodeGroup` also useless.
+
+
+###### Pool-3:5:1:1:2 (min:max:maxsurge:maxunavail:numzones)
+```
+MachineDeployment_Z0(Minimum=2|Maximum:3|MaxSurge:1|MaxUnavailable:1)
+MachineDeployment_Z1(Minimum=1|Maximum:2|MaxSurge:0|MaxUnavailable:0)
+```
+Consider the case above - `MaxSurge`  for `Z1` is computed as `0`
+
+The above cases are not un-common. Thre are `~65` Worker Pools in LIVE Landscape with `0 < WorkerPoolMax < WorkerPoolZones`
+
+
+
+
+### Deficiencies of Existing Solution
+
+1. The current static node group distribution skews distribution towards earlier specified zones over later specified zones.
+2. If zone length exceeds worker pool max, then later node groups get zero values effectively becoming useless.
+3. If the earlier node groups have quota issues, then later node groups have reduced capacity and cannot meet the specified availability. 
+
+
+### Goals
+
+1. We remove the current static node group assignment and move to a dynamic node group policy leveraging [node-group-auto-discovery](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/main.go#L156)
+2. We change our MCM [CloudProvider](https://github.com/gardener/autoscaler/blob/053c0d5176cb2d195e3baf333b05ceea99eedb58/cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go#L52) implementaion to support dynamic node groups and node group auto-discovery.
+3. We change the deployment of the CA Deployer to construct the command string differently.
+4. We fix invalid shoot clusters on all landscapes where `WorkerPool.Maxium >= WorkerPool.NumZones` 
```

Comment:
Is the comparison sign correct? I would rather expect `Maximum < Zones` to be the issue.

Reply:
oops